src/Utilities/WellKnownTypeProvider.cs

```diff
@@ -67,6 +73,36 @@ private WellKnownTypeProvider(Compilation compilation)
         /// </summary>
         public INamedTypeSymbol GenericIEquatable { get; }
 
+        /// <summary>
+        /// Mapping of full name to <see cref="INamedTypeSymbol"/>.
+        /// </summary>
+        private ConcurrentDictionary<string, INamedTypeSymbol> FullNameToType { get; }
+
+        /// <summary>
+        /// Attempts to get the type by the full type name.
+        /// </summary>
+        /// <param name="fullTypeName">Namespace + type name, e.g. "System.Exception".</param>
+        /// <param name="namedTypeSymbol">Named type symbol, if any.</param>
+        /// <returns>True if found in the compilation, false otherwise.</returns>
+        public bool TryGetTypeByMetadataName(string fullTypeName, out INamedTypeSymbol namedTypeSymbol)
+        {
+            if (!FullNameToType.TryGetValue(fullTypeName, out namedTypeSymbol))
+            {
+                namedTypeSymbol = Compilation.GetTypeByMetadataName(fullTypeName);
+
+                // Even if the compilation gives back null, still cache the null to avoid future lookups.
+                FullNameToType.TryAdd(fullTypeName, namedTypeSymbol);
+            }
+
+            return namedTypeSymbol != null;
```

Comment:
Can this be simplified using [GetOrAdd](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.getoradd?view=netframework-4.7.2)?